pkg/metrics/observer/query_observer.go

```diff
@@ -0,0 +1,74 @@
+package observer
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	prometheusv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	errorutil "k8s.io/apimachinery/pkg/util/errors"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	routev1client "github.com/openshift/client-go/route/clientset/versioned/typed/route/v1"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	metricsclient "github.com/openshift/library-go/pkg/metrics/client"
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+var defaultPollingDuration = 30 * time.Second
+
+type QueryHandlerFunc = func(ctx context.Context, recorder events.Recorder, currentValue string, valueType string) error
+
+// QueryHandler represents a single Prometheus query and a handler function that should run on result of given query.
+type QueryHandler struct {
+	Query   string
+	Handler QueryHandlerFunc
+}
+
+type prometheusMetricObserver struct {
+	prometheusClient prometheusv1.API
+	queryHandlers    []QueryHandler
+}
+
+// NewPrometheusMetricObserver allows to continuously observe particular Prometheus query result value and react to it.
+// The name should be unique name for this observer.
+// The queryHandlers represents list of Prometheus queries and handlers where the result value and type are passed.
+// The clients are needed to initialize that prometheus API client.
+func NewPrometheusMetricObserver(name string, queryHandlers []QueryHandler, serviceClient corev1client.ServicesGetter, secretsClient corev1client.SecretsGetter, configsClient corev1client.ConfigMapsGetter,
+	routeClient routev1client.RoutesGetter, recorder events.Recorder) (factory.Controller, error) {
+	observer := &prometheusMetricObserver{
+		queryHandlers: queryHandlers,
+	}
+	prometheusClient, err := metricsclient.NewPrometheusClient(serviceClient, secretsClient, configsClient, routeClient)
+	if err != nil {
+		return nil, err
+	}
+	observer.prometheusClient = prometheusClient
+	return factory.New().ResyncEvery(defaultPollingDuration).WithSync(observer.sync).ToController(name, recorder.WithComponentSuffix(name)), nil
+}
+
+func (o *prometheusMetricObserver) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	var errors []error
+	for _, handler := range o.queryHandlers {
+		result, warnings, err := o.prometheusClient.QueryRange(ctx, handler.Query, prometheusv1.Range{
+			Start: time.Now().Add(-defaultPollingDuration),
+			End:   time.Now(),
+		})
+		if err != nil {
```

Comment:
@deads2k @hexfusion i wonder if we should set degraded conditions here if we are unable to query prometheus for metrics (this might cause some churn during bootstrap?)